Full Touch support for default Marlin Menus - Touch to Select!

rhapsodyv opened the pull request:
### Description

Hi,

Using a touch screen with default Marlin menu is a bit annoying because we need to use the buttons bellow the screen.
But it changes now :-)

I'm enabling it with TOUCH_BUTTONS. I can emulate a click too, but seems touch to select is better.

The change could be done just in the xpt2046.cpp file, but I think the "screen_click" function must be part of MarlinUI... Tried to keep the thinks in the right place.

### Benefits

It's more easier and user friendly to touch directly where you want, in a TFT display that uses the default marlin menu.

Some users have touch displays that don't have any other UI. Link TronXY users using Marlin. So it helps this users to have a better UI experience.

This videos explains better :-)


![ezgif-4-208ce3332154](https://user-images.githubusercontent.com/81722/84099661-835e8680-a9e0-11ea-9709-924d5d81cae0.gif)
![ezgif-4-b166bfb54b8c](https://user-images.githubusercontent.com/81722/84099958-34652100-a9e1-11ea-9bd8-e22f49fa8c26.gif)

**I think I must explain my code decisions in the PR:**

I could do it a lot better, with more classes and stuff, supporting double-touch, touch on icons, edit screen, and so on.

But, my main goal here was: 

_keep changes at minimum, dont break any compatibility, yet supporting the most used touch actions for the users, to get this PR accepted as soon as possible without any risk to Marlin_

So, with only 3 files and a few lines changed, we get the most action covered:
   - touch to select menus
   - touch to select confirm action
   - touch to go to next/prev pages

Why do I want to keep things so minimum now? Because the TFT users that don't have smart screens need it, as the case of tronxy x5sa users - we had working on this about a week or so.

I already working in a 2.0 version of this full touch, that each screen could define its touchable actions - icons, menus, edit, support double touch, etc. The code is better, but the change is a lot bigger. If I did the big version first, the tronxy x5sa users (and a lot of others users) will need to wait a long time to use it, because a more complex and big PR isnt so easy to merge without break anything - it needs more review, more tests, and so on...

Well, I think I could explain my design/code decisions to help it get accepted as soon as possible :-)
I have a few users testing it, and they are enjoying a lot!

Thanks!

rhapsodyv committed:
It enables a FULL touch support using default Marlin menus. That is: the user can touch DIRECTLY on the menu he wants to select!!

a) Touch to select menus.
b) In confirmation screen, lelt screen select left btn, right screen select right btn

rhapsodyv committed:
forgot to remove on merge

rhapsodyv committed:
forgot to remove on merge

thinkyhead commented:
Nice, this is a very welcome feature. I'm also working on making the native Marlin menu system work on DWIN screens with a standard DWIN_SET, so it will be great to have touch support for that from the start.

thinkyhead committed:
const

rhapsodyv commented:
> Nice, this is a very welcome feature. I'm also working on making the native Marlin menu system work on DWIN screens with a standard DWIN_SET, so it will be great to have touch support for that from the start.

We also can make the status screen act if it has buttons. For example: touch on the cooler icon goes to fan speed control screen; touch on nozzle temperature goes to nozzle temp config screen, and so on..

thinkyhead committed:
Update ultralcd.cpp

thinkyhead committed:
Update xpt2046.cpp

thinkyhead committed:
Update xpt2046.cpp

thinkyhead committed:
Update xpt2046.cpp

thinkyhead commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -1438,6 +1438,22 @@ void MarlinUI::update() {
+      }
+      //confirm and other windows, right screen is +, left screen is -
+      else {
+        encoderDiff = row * ENCODER_PULSES_PER_STEP;
```
I believe this doesn't need to include `row` in the calculation. One "row" worth of pulses is enough to indicate the encoder step.

thinkyhead replied:
I see that you could use the "row" as a way to increment or decrement by a larger value. I'm not sure that users will be able to see this nuance, and they might think there is a malfunction.

rhapsodyv replied:
When screen has itens (one per line), a use row to find the current steps needed to set the position touched by the user.

When screen has no itens (like confirm screen), I use col to know if user touched left or right.

rhapsodyv replied:
The two "modes" of touch are exactly in the gifs. I was thinking about one for edit screens, but didn't work yet.
And, of course, we can have on mode for direct touch on target areas, like touch on icons in the status screen. (dint yet too).

thinkyhead committed:
Only inc/dec by 1 (not proportional to "row")

thinkyhead commented:
Is this touch button device only able to report touch events and not un-touch events? If the device could tell us that the touch is being sustained, we could use that data to continually jog the edited value, with acceleration, until it is released.

thinkyhead commented on Marlin/src/feature/touch/xpt2046.cpp:
```diff
@@ -77,14 +81,24 @@ uint8_t XPT2046::read_buttons() {
+
+  // Send the touch to the UI (which will simulate the encoder wheel)
+  MarlinUI::screen_click(row, col, x, y);
+  return 0;
```
What happens if you return `EN_C` here instead of 0? Does the correct menu item get activated? If not, we can fix that!

rhapsodyv replied:
Do you want simulate the click? I didnt try, but seems will work. But I liked more the touch acting like selection.

thinkyhead replied:
But if the touch is on the already-selected item…

rhapsodyv replied:
It will be nice. But, in this case, who needs to decide if clicks or not is the UI part, right? Right there we can check if the current line is the same one touched and send a click.

I will work on that in a minute.

tpruvot replied:
the code already do multiple "clicks" if you move the finger and stay pressed...

tpruvot commented:
there is a few small issues, but the idea is good : https://drive.google.com/file/d/1oIJs7qwAxI1L76X1XXeGwSEdX2_y98QZ/view?usp=sharing

(gdrive could take a few minutes to process the upload of the vid)

rhapsodyv commented:
> there is a few small issues, but the idea is good : https://drive.google.com/file/d/1oIJs7qwAxI1L76X1XXeGwSEdX2_y98QZ/view?usp=sharing
> 
> (gdrive could take a few minutes to process the upload of the vid)
> 
> * on status screen, edit the feedrate... maybe should avoid by default...
> * menu y scroll behavior seems reversed

Seems the video isnt public.

What display are you using?

tpruvot commented:
ILI9341 (U20 config in marlin samples)

thinkyhead commented:
> public but in process

The page asks visitors to provide their email address for an access link.

rhapsodyv commented:
> ILI9341 (U20 config in marlin samples) video is public but in process, take some time on google drive

rhapsodyv commented:
clicked Hahahaha wrong button!! "close comment"!!! dammit

rhapsodyv commented:
clicked Hahahaha wrong button!! "close comment"!!! dammit

tpruvot commented:
yep, its only 25MB but discord only allow 8... cant share it there rmm.. let me see where i can upload...

rhapsodyv commented:
> yep, its only 25MB but discord only allow 8... cant share it there rmm.. let me see where i can upload...

The bottom buttons are working correctly, is just the menu touch that is wrong? it's inverted?

tpruvot commented:
yes... y seems correct else buttons would not work...

tpruvot commented on Marlin/src/feature/touch/xpt2046.cpp:
```diff
@@ -77,14 +81,24 @@ uint8_t XPT2046::read_buttons() {
+  int8_t col = (x % (LCD_FULL_PIXEL_WIDTH) ) * (LCD_WIDTH ) / (LCD_FULL_PIXEL_WIDTH),
+         row = (y % (LCD_FULL_PIXEL_HEIGHT)) * (LCD_HEIGHT) / (BUTTON_AREA_TOP);
+
+  row = (LCD_HEIGHT) - row - 1; // TODO: Can LCD or sensor be inverted?
```
i guess the reversed Y in (my) menus is caused there... also, we have 32 extra pixels on top, i need to press in this extra space to select the last menu item...

rhapsodyv replied:
In theory, XPT2046_X_CALIBRATION would solve inversion issues, but I have only one display. Can you add to chat about it? https://www.facebook.com/victormateus.oliveira/

tpruvot replied:
i need to double check ive the last version of this pull request, for some reason github PR page doesnt show the same as what i merged from your branch (im on discord tpruvot)

tpruvot replied:
arf, Tanguy on discord, X is ok, only the row is reversed... in menus

rhapsodyv replied:
mine: rhapsodyv#9981 on discord

tpruvot replied:
so yes, this line 97 should be fully commented... there is no logic reason to inverse the row... x and y values are supposed calibrated at this point, with top left = 0 0

google drive video is now processed/shared

tpruvot commented on Marlin/src/lcd/ultralcd.cpp:
```diff
@@ -1438,6 +1438,22 @@ void MarlinUI::update() {
+    void MarlinUI::screen_click(const uint8_t row, const uint8_t col, const uint8_t x, const uint8_t y) {
+      if (screen_items > 0)
+        encoderDiff = ENCODER_PULSES_PER_STEP * (row - encoderPosition + encoderTopLine);
+      else
```
i would replace this else by "else if (!on_status_screen()) to avoid feedrate change by mistake.... or/and maybe try to trigger icon's clicks ...

rhapsodyv committed:
fixed row inverstion (was a calibration issue). Thanks to @tpruvot. The current touch assumes the screen is 320x240, not the LCD_FULL*, that can be upscaled.

rhapsodyv committed:
by now, status screen dont have anything touchable. Thanks to @tpruvot pointing out.

rhapsodyv committed:
removed extra space

rhapsodyv committed:
limiting x and y; modulo isnt needed

rhapsodyv commented:
changes done:
- fixed invert issue
- removed modulo and limited x and y
- ignore touch on status screen

I think we have a first version.

@thinkyhead I dont know if its a good idea use touch as a click for already selected because the screen sized. Could affect the user experience entering where they dont want.

Future work:
 - The screen could define touchable areas with actions to execute - Example: Confirm can define yes/no buttons to be touchable, so when click, it already exec the action; Status screen have a lot of potencial to goes directly to the config when touch in some areas, etc
- Editable screens could have a special treatment. But right now we can't know which type of screen the user is on. We need some enum, like: screen_type { SCREEN_MENU, SCREEN_STATUS, SCREEN_CONFIRM, SCREEN_EDIT, ... }...

jmz52 commented:
@rhapsodyv, you might be interested in new implementation of TFT and touch screen support.
It uses full screen, has better fonts and graphics and touch interaction with displayed elements.
Support TFT screens with SPI and FSMC interfaces on STM32F1 and STM32F4 MCUs.
Pictures - https://github.com/MarlinFirmware/Marlin/issues/18129
Code - https://github.com/MarlinFirmware/Marlin/pull/18130

rhapsodyv committed:
On menu screens, right side (last 3 cols) acts like a scroll - half up => prev page, half down = next page

rhapsodyv committed:
Fixed the row calc to consider LCD_PIXEL_OFFSET_Y

rhapsodyv commented:
@tpruvot take a look in the last commit. I have fixed the Y offset calc. I have tested with a big and small LCD_PIXEL_OFFSET_Y.
Can you test in your screen?

le3tspeak commented:
Is it possible to extend this by the following that if you want to change values eg the temperature of the nozzle when I tap the screen on the right for plus and left for minus.

And there is also an error at the moment where you can scroll in a menu, for example when changing the Nozzle temperature menu point it goes to the MAX value when I move my finger up and when I go down it is 0

rhapsodyv commented:
> Is it possible to extend this by the following that if you want to change values eg the temperature of the nozzle when I tap the screen on the right for plus and left for minus.
> 
> And there is also an error at the moment where you can scroll in a menu, for example when changing the Nozzle temperature menu point it goes to the MAX value when I move my finger up and when I go down it is 0

Yes, sure. Right now I didnt any handle for the edit screens. I need a way to identify it, but I dint want to make a lot of changes in the code. By now, just works:
 - menu select
 - confirm select
 - next/prev page in menus

I will try to find a way to identify edit screens without changing a lot of places. In my local repo, I even did the the confirmation screen buttons touchable. 
But I think it will be for a new PR.

@thinkyhead what do you think?

rhapsodyv commented:
I think I must explain my code decisions in the PR:

I could do it a lot better, with more classes and stuff, supporting double-touch, touch on icons, edit screen, and so on.

But, my main goal here was: 

_keep changes at minimum, dont break any compatibility, yet supporting the most used touch actions for the users, to get this PR accepted as soon as possible without any risk to Marlin_

So, with only 3 files and a few lines changed, we get the most action covered:
   - touch to select menus
   - touch to select confirm action
   - touch to go to next/prev pages

Why do I want to keep things so minimum now? Because the TFT users that don't have smart screens need it, as the case of tronxy x5sa users - we had working on this about a week or so.

I already working in a 2.0 version of this full touch, that each screen could define its touchable actions - icons, menus, edit, support double touch, etc. The code is better, but the change is a lot bigger. If I did the big version first, the tronxy x5sa users (and a lot of others users) will need to wait a long time to use it, because a more complex and big PR isnt so easy to merge without break anything - it needs more review, more tests, and so on...

Well, I think I could explain my design/code decisions to help it get accepted as soon as possible :-)
I have a few users testing it, and they are enjoying a lot!

Thanks!

thinkyhead commented:
> I will try to find a way to identify edit screens without changing a lot of places.

Add a `bool` flag called `on_edit_screen` to MarlinUI. Set this flag to `true` in `MenuEditItemBase::goto_edit_screen` and set it to `false` in `MarlinUI::goto_previous_screen`.

thinkyhead committed:
Fix Y range check

tpruvot commented:
changes are good, just a visual/design issue to fix with #18273 

edit value touch is not yet really "usable" they tend to only do min/max values, but well, can help too

dandantsui commented:
Just tested this with my MKS_Robin with Robin_TFT28 and it is working well! The scrolling on the right side is very ingenious! Maybe it could have an icon or scrollbar of some sort?

rhapsodyv commented:
> Just tested this with my MKS_Robin with Robin_TFT28 and it is working well! The scrolling on the right side is very ingenious! Maybe it could have an icon or scrollbar of some sort?

Good to know! I'm working on better scroll and "touch space" for smaller screens (to fit the fingers!!).

But I just will send a PR when the smaller screen I ordered arrive. Currently I just have one 3.5" 480x320.

Some preview 
Before:
![image0](https://user-images.githubusercontent.com/81722/84548017-1cd8b180-acdb-11ea-810c-1b3a8b8a9705.jpg)

After:
![image0-2](https://user-images.githubusercontent.com/81722/84548021-206c3880-acdb-11ea-88d6-101cd818afe6.jpg)

dandantsui commented:
Oops! I meant to say TFT32! Just an upscaled version of TFT28. Not sure why my stepper motors aren’t moving with the bugfix branch. Will have to resolve that before I can continue.

dandantsui commented:
Hmm steppers are working now. But have a glitchy TMC connection (Some axis saying no connection or all low, randomly). I think something was changed in bugfix that is causing intermittent single wire communication on TMC2209. Switching back to 2.0.5.3 with the same settings shows no problems.  I will occasionally flash back to bugfix to test things out if you’d like!

thinkyhead commented:
> glitchy TMC connection

If a bug report is filed with sufficient information then someone will look into it.

dandantsui commented:
@thinkyhead I’m really not sure what’s going on. The connection errors are very sporadic and randomly jump between the drivers. There is no set driver that reports connection error but rather random amount return that problem.

Is there any way I can provide more info besides the sporadic M122 response? I am more than happy to help diagnose this out.
